pkg/authorizer/authorizer.go

```diff
@@ -0,0 +1,144 @@
+package authorizer
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/3scale/3scale-go-client/threescale/api"
+	"github.com/3scale/3scale-istio-adapter/pkg/threescale/authorizer"
+	"github.com/3scale/3scale-porta-go-client/client"
+	log "github.com/sirupsen/logrus"
+)
+
+type Authorizer struct {
+	stopChan chan struct{}
+	config   Config
+	manager  *authorizer.Manager
+}
+
+type Config struct {
+	serviceAPICacheConfig *authorizer.SystemCacheConfig
+	authorizationAPICache bool
+	adminPortalURL        *url.URL
+	environment           string
+	serviceID             string
+}
+
+// TODO: Add a default builder that uses cache by default with everything set to sane defaults.
+// TODO: Add a more customizable builder.
+
+func New(AdminPortalURL *url.URL, environment, serviceID string, ServiceAPICache,
+	AuthorizationAPICache bool) (*Authorizer, error) {
+	var httpClient *http.Client
+	var systemCacheConfig authorizer.SystemCacheConfig
+
+	if serviceID == "" {
+		return nil, fmt.Errorf("missing serviceID")
+	}
+	if environment == "" {
+		log.Warning("Environment variable is empty, running in sandbox mode.")
+		environment = "sandbox"
+	}
+
+	//TODO: Match defaults with 3scale-istio-adapter lib
+	if ServiceAPICache {
+		systemCacheConfig.MaxSize = 100
+		systemCacheConfig.NumRetryFailedRefresh = 3
+		systemCacheConfig.RefreshInterval = 30 * time.Second
+		systemCacheConfig.TTL = 120 * time.Second
+	}
+
+	stop := make(chan struct{})
+
+	systemCache := authorizer.NewSystemCache(systemCacheConfig, stop)
+	builder := authorizer.NewClientBuilder(httpClient)
+	manager, err := authorizer.NewManager(builder, systemCache, AuthorizationAPICache)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Authorizer{
+		stopChan: stop,
+		manager:  manager,
+		config: Config{
+			serviceAPICacheConfig: &systemCacheConfig,
+			authorizationAPICache: AuthorizationAPICache,
+			adminPortalURL:        AdminPortalURL,
+			environment:           environment,
+			serviceID:             serviceID,
+		},
+	}, nil
+}
+
+func (a *Authorizer) Stop() {
+	a.stopChan <- struct{}{}
+}
+
+func (a *Authorizer) AuthzRep(request *http.Request) (isAuthorized bool, err error) {
```

Comment:
What do you think here about having `AuthzRep` take a package type like `Request` or something, then we could add a layer or some functions to do conversions. So `*http.Request --> authorizer.Request`, in this way it would allow us to use the authoriser for many purposes, not just http but we could do a conversion say in the istio adapter to turn the gRPC request into a `Request` and yet still use the same `AuthzRep` functionality that would be well tested.

Doing this would allow us to scale out to any other ideas or integrations that might come up by writing a single conversion function, wdyt?